chains/ethereum/server/src/block_provider.rs

```diff
@@ -0,0 +1,368 @@
+use super::client::BlockFinalityStrategy;
+use std::{
+    fmt::{Debug, Display},
+    sync::Arc,
+};
+use tokio::{
+    sync::Mutex,
+    time::{Duration, Instant},
+};
+
+use crate::utils::PartialBlock;
+use futures_util::{future::BoxFuture, Future, FutureExt};
+use rosetta_config_ethereum::{
+    ext::types::{crypto::DefaultCrypto, BlockIdentifier},
+    AtBlock,
+};
+use rosetta_ethereum_backend::EthereumRpc;
+
+/// Block Provider
+pub trait BlockProvider: Unpin {
+    /// Error type
+    type Error: Unpin + Send + Sync + 'static;
+    /// Future type
+    type BlockAtFut: Future<Output = Result<Option<Arc<PartialBlock>>, Self::Error>>
+        + Unpin
+        + Send
+        + 'static;
+    /// Future type
+    type LatestFut: Future<Output = Result<Arc<PartialBlock>, Self::Error>> + Unpin + Send + 'static;
```

Comment:
Yeah this is not needed anymore, It was needed before because one struct needed to know the explicit future type returned by each method, for example when using [futures-utils](https://docs.rs/futures-util/latest/futures_util/future/index.html) to compose futures, sometimes is necessary do something like this:
```rust
use futures_utils::{stream::FuturesUnordered, futures::Either};
use futures_timer::Delay;

type FutureOrTimeout<Fut> = Either<Fut, Delay>;

struct SomeHelper<T: BlockProvider> {
    pending: FuturesUnordered<FutureOrTimeout<T::LatestFut>>,
}
```

Reply:
Btw this is an style for programming concurrent systems in Rust I learned from rust-libp2p:
- https://github.com/libp2p/rust-libp2p/blob/v0.51.0/protocols/ping/src/handler.rs#L256-L263
- https://github.com/libp2p/rust-libp2p/blob/v0.51.0/swarm/src/handler/multi.rs#L225-L231

Instead of boxing everything by using `BoxFuture<..>`, all futures are explicitly typed, this help to debug and jump to the future definition, it also allow you to do pattern matching or calls methods to change the future behavior, like when using an [Abortable](https://docs.rs/futures-util/latest/futures_util/future/struct.Abortable.html) future, for some operation that needs to be explicitly aborted (like websocket subscriptions which need to be explicitly unsubscribed, aka do another async request).

Reply:
Bug I agree this style is very verbose and less redeable, is only necessary when you have some advanced features, example:
- I use it a lot in`rosetta-server/src/ws/reconnect_impl.rs`, to abstract from the timechain the websocket retry connection mechanism.
- Here I need it for the **BlockProvider** to ensure only one thread is requesting the `Finalized` or `Latest` block at time, a mutex guarantee that by locking a **Mutex** before request the Finalized or Latest block.